### ISSUE TITLE
Improve ETD Import performance?

### ISSUE THREAD

**Issue by mark-dce** [curationexperts/alexandria-v2#443]
**RECORDS 234-238** average 145.5 sec/record
```
2015-08-11 06:34:11 +0000  Created etd f3/sj/1h/hr/f3sj1hhr (003891600)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 06:36:33 +0000  Created etd f3/ns/0r/s1/f3ns0rs1 (003891601)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 06:39:00 +0000  Created etd f3/j1/01/3v/f3j1013v (003891602)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 06:41:26 +0000  Created etd f3/db/7z/r2/f3db7zr2 (003891603)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 06:43:53 +0000  Created etd f3/8k/77/14/f38k7714 (003891604)
```

**RECORDS 343-347** average 274.5 sec/record
```
2015-08-11 12:44:10 +0000  Created etd f3/93/0r/32/f3930r32 (003914789)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 12:48:45 +0000  Created etd f3/5b/00/cq/f35b00cq (003914790)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 12:53:17 +0000  Created etd f3/1j/97/pw/f31j97pw (003914791)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 12:57:49 +0000  Created etd f3/ws/8r/6m/f3ws8r6m (003914792)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 13:02:28 +0000  Created etd f3/s4/6p/wh/f3s46pwh (003914793)
```

*NOTE* I added `Time.now` to https://github.com/curationexperts/alexandria-v2/blob/master/lib/importer/factory/etd_factory.rb#L39

It looks like traject might be reloading the entire rails environment for each record. (or something weird is going on).  The excerpt below is from a single import in a screen session - I didn't redeploy code and had added the timestamp manually before starting the import (I thought I'd put it in the wrong place because it wasn't showing when I started)

```
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
  Updated etd f3/4t/6g/gn/f34t6ggn (003891596)  
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
  Updated etd f3/5t/3h/fs/f35t3hfs (003891597)  
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
  Updated etd f3/22/2r/qc/f3222rqc (003891598)  
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
  Updated etd f3/x9/28/7g/f3x9287g (003891599)  
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 06:34:11 +0000  Created etd f3/sj/1h/hr/f3sj1hhr (003891600)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 06:36:33 +0000  Created etd f3/ns/0r/s1/f3ns0rs1 (003891601)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 06:39:00 +0000  Created etd f3/j1/01/3v/f3j1013v (003891602)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 06:41:26 +0000  Created etd f3/db/7z/r2/f3db7zr2 (003891603)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 06:43:53 +0000  Created etd f3/8k/77/14/f38k7714 (003891604)
  Updated collection fk/4g/x4/hm/fk4gx4hm1c (etds)
2015-08-11 06:46:22 +0000  Created etd f3/mc/8x/45/f3mc8x45 (003891605)
```

**Comment by jcoyne** [comment]
The ETD collection has to be loaded on every save.  It doesn't take too long to fetch from Fedora:
  HTTP GET (31.23ms)   http://127.0.0.1:8983/fedora/rest/dev/fk/4s/q9/39/fk4sq9391j

But it takes  a very long time to parse it:
  Load LDP (1572.5ms) http://127.0.0.1:8983/fedora/rest/dev/fk/4s/q9/39/fk4sq9391j

**Comment by jcoyne** [comment]
Currently it seems to be CPU bound (ruby running at 100% most of the time).  Currently we can't run this job multithreaded (concurrent modification of the collection), so my next suggestion is for the importer not to add the ETD to a collection on import.  Then the import can be run on several CPUs.   At the end of importing ETDs, there should be a job that queries Solr for the Ids of every ETD and adds them to the collection.

**Comment by jcoyne** [comment]
Fixing https://github.com/projecthydra/active_fedora/issues/875 is another optimization.

**Comment by jcoyne** [comment]
Superseded by #459
